lib/pbench/server/database/models/api_keys.py

```diff
@@ -20,40 +23,60 @@ def __str__(self):
         return repr(self.message)
 
 
-class APIKeys(Database.Base):
+class APIKey(Database.Base):
     """Model for storing the API key associated with a user."""
 
     __tablename__ = "api_keys"
-    api_key = Column(
-        String(500), primary_key=True, unique=True, nullable=False, index=True
-    )
+    api_key = Column(String(500), primary_key=True)
     created = Column(TZDateTime, nullable=False, default=TZDateTime.current_time)
     # ID of the owning user
     user_id = Column(String, ForeignKey("users.id"), nullable=False)
 
     # Indirect reference to the owning User record
     user = relationship("User")
 
+    def __str__(self):
+        return f"key: {self.api_key}"
+
+    class DuplicateValue(commonDBException):
+        """Attempt to commit a duplicate unique value."""
+
+        def __init__(self, api_key: "APIKey", cause: str):
+            self.api_key = api_key
+            self.cause = cause
+
+        def __str__(self) -> str:
+            return f"Duplicate api_key {self.api_key}: {self.cause}"
+
+    class NullKey(commonDBException):
+        """Attempt to commit an APIkey row with an empty required column."""
+
+        def __init__(self, api_key: "APIKey", cause: str):
+            self.api_key = api_key
+            self.cause = cause
+
+        def __str__(self) -> str:
+            return f"Missing required key in {self.api_key.as_json()}: {self.cause}"
+
     def add(self):
         """Add an api_key object to the database."""
         try:
             Database.db_session.add(self)
             Database.db_session.commit()
         except Exception as e:
             Database.db_session.rollback()
-            self.logger.error("Can't add {} to DB: {}", str(self), str(e))
-            raise APIKeyError("Error adding api_key to db")
+            self.logger.error("Can't add key:{} to DB: {}", str(self), str(e))
+            raise commonDBException._decode(self, e) from e
```

Comment:
Note that there are one or two other issues here:  if we cannot decode the exception and raise our own "duplicate key" or "null key" error, then we probably _don't_ want to reraise the original exception here, since it won't be an `APIKeyError` (or `commonDBException`) derivative; instead we presumably want to do something similar to what you were doing before and wrap the SQL exception (or, at least, its message) in an `APIKeyError`.  (The other issue is that `APIKeyError` and `commonDBException` don't have a common parent class (until you get all the way up to `Exception`), which means that the caller cannot catch them both without either catching them separately or catching all `Exception`'s...which is disappointing; hopefully, your rework of `_decode()` will also rectify the object hierarchies.)